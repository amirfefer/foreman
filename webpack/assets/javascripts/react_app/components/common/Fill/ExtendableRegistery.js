const components = {};

export const add = (SlotId, fillId, component, weight, overrideProps) => {
  if (components[SlotId] === undefined) {
    components[SlotId] = [];
  }

  component = component || overrideProps;

  components[SlotId].push({ fillId, component, weight });
};

export const remove = (SlotId, fillId) => {
  const slotItems = components[SlotId];

  components[SlotId] = slotItems.filter(item => item.fillId !== fillId);
};

export const getComponentsBySlotId = id => components[id];
