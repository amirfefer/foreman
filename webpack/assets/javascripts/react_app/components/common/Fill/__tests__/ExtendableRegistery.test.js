import React from 'react';
import { add, remove, getComponentsBySlotId } from '../ExtendableRegistery';

jest.unmock('lodash');
jest.unmock('../ExtendableRegistery');

const componnentOne = () => <div> Component 1</div>;
const componnentTwo = () => <div> Component 2</div>;
const props = { someProp: true };

describe('Extendable Registery', () => {
  it('should render two components by weights', () => {
    add('slot-id', 'fill-id-1', componnentOne, 100);
    add('slot-id', 'fill-id-2', componnentTwo, 200);
    add('another-slot-id', 'fill-id', props, 300);

    expect(getComponentsBySlotId('slot-id')).toMatchSnapshot();
  });
  it('should render one component after a removal', () => {
    remove('slot-id', 'fill-id-2');
    expect(getComponentsBySlotId('slot-id')).toMatchSnapshot();
  });
});
