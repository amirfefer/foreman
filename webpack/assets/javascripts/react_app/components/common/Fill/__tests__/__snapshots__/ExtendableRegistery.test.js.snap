// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extendable Registery should render one component after a removal 1`] = `
Array [
  [Function],
]
`;

exports[`Extendable Registery should render one component after a removal 2`] = `[Function]`;

exports[`Extendable Registery should render two components by weights 1`] = `[Function]`;

exports[`Extendable Registery should render two components by weights 2`] = `
Array [
  [Function],
  [Function],
]
`;
